OCLOMRS-509: Edit a concept answer in Q-A concepts without having to remove it

danuluma opened the pull request:
# JIRA TICKET NAME:
[ Edit a concept answer in Q-A concepts without having to remove it](https://issues.openmrs.org/browse/OCLOMRS-509)

# Summary:
Currently, if one wishes to change an answer to a concept, they have to remove it and then add a new answer. An ideal situation would be having the ability to edit a concept directly by clicking in the concept field.

danuluma commented:
There are currently no conflicts with the master branch hence a rebase would be unneccessary

Karuhanga commented on src/components/dictionaryConcepts/containers/SelectAnswers.jsx:
```diff
@@ -13,8 +13,24 @@ class SelectAnswers extends Component {
+    this.setState({ inputValue: this.props.defaultValue });
+  }
+
+    resetInput = (e) => {
```
This approach to deleting the answer introduces a bug. Since both the `Add` and `Edit` Concept share the same `Answer` component, it expects `removeCurrentAnswer` to be passed as a prop, which it is not in the add concept case.
It therefore crashes when one places the cursor in the input box for creating a new answer.
[Recording](https://drive.google.com/file/d/1ISGperfma3cBujGjXAui0WSLo3sB-l83/view?usp=sharing).
Screenshot:
<img width="989" alt="image" src="https://user-images.githubusercontent.com/12200860/56604600-040ac980-660b-11e9-8172-676a246a21f0.png">

danuluma replied:
Thanks for helping discovering this. I'm working on it

danuluma replied:
Fixed

danuluma commented on src/tests/dictionaryConcepts/actions/dictionaryConcept.test.js:
```diff
@@ -725,6 +729,50 @@ describe('Testing Edit concept actions ', () => {
+
+    const store = mockStore(mockConceptStore);
+    return store.dispatch(backgroundRemoveConceptMapping(data)).then().catch((error) => {
+      expect(error).toBeTruthy();
```
It does. I'm responding to the mocked request with status 500. Any status outside 2XX is assumed to be a failed request hence the catch block is called.

Karuhanga replied:
Unfortunately, when I pulled the changes and run this test, it passed regardless of what is inside the catch block. I think that would be a good check for whether or not those lines are being hit.

danuluma replied:
Done

danuluma commented on src/components/dictionaryConcepts/containers/EditConcept.jsx:
```diff
@@ -483,6 +490,16 @@ export class EditConcept extends Component {
+    const data = {
+      references: [info.answerUrl],
+    };
+    await backgroundRemoveConceptMap(data);
```
Let me work on that too

danuluma replied:
Fixed

danuluma commented:
> Good work on this @danuluma.
> 
> #### A few observations;
> * In my opinion the approach this PR takes to updating an answer may need some rework. In general, it feels un-intuitive to a user to immediately delete a mapping once they click it. I think the expected behavior would be to perform this action when update is clicked.
> * The concepts displayed after searching do not seem to reflect the search query. Please clarify in regards to this behavior.
> 
> #### A few more observations as well;

Thanks @Karuhanga. Agreed, that would have been an alternative approach to this. I however took a different implementation of the same

Noted. However, this PR didn't touch on the search functionality. It still works as before. We might need to create a separate PR if there's any bug in that.

dkayiwa commented on src/components/dictionaryConcepts/components/helperFunction.js:
```diff
@@ -70,3 +70,7 @@ export const ATTRIBUTE_NAME_SOURCE = 'source';
 export const KEY_CODE_FOR_ESCAPE = 27;
 export const KEY_CODE_FOR_SPACE = 32;
+export const mapSources = {
+  'MLG': 'Mulago Hospital (Dictionary)',
```
Is this only for testing purposes?

danuluma replied:
Hi Daniel, I have used it in `src/components/dictionaryConcepts/components/AnswerRow.jsx` Instead of having to hard code the sources and a conditional statement in that file, I created a dictionary of the possible sources in this file, imported it to the `AnswerRow` and then simply fetch corresponding value of the source name.

dkayiwa replied:
Is Mulago Hospital also a source to be used by Andy? I thought by doing this you are already hard coding.

danuluma replied:
Okay I now get it, let me change that

danuluma replied:
I've changed it to display the `currentDictionary` - which is the dictionary containing the concept being edited, as it was before. Is it okay?

dkayiwa commented on src/components/dictionaryConcepts/containers/EditConcept.jsx:
```diff
@@ -483,6 +490,20 @@ export class EditConcept extends Component {
+      this.removeUnsavedMappingRow(info.answerUrl);
+      unPopulateAnswer(info.answer);
+    } catch (error) {
+      notify.show('An error occurred, can\'t edit this answer', 'error', 3000);
```
Which kind of error is this?

danuluma replied:
This notifies of any error in case the mapping fails to be removed. I've updated the message.

dkayiwa commented on src/redux/actions/dictionaries/dictionaryActionCreators.js:
```diff
@@ -147,6 +147,21 @@ export const removeConceptMapping = (data, source) => dispatch => {
+          dispatch(fetchDictionaryConcepts(ConceptsToFetch));
+      }
+    )
+    .catch((error) => {
```
Will this be always a network error?

danuluma replied:
The network error should only be triggered if there's no response from the api.

dkayiwa commented on src/redux/actions/types.js:
```diff
@@ -72,3 +72,4 @@ export const SET_PERVIOUS_PAGE = '[page] prev_page';
 export const SET_NEXT_PAGE = '[page] next_page';
 export const SET_CURRENT_PAGE = '[page] current_page';
 export const IS_LOADING = '[ui] loader_spinning';
+export const UN_POPULATE_THIS_ANSWER = '[concept] unpopulate an answer';
```
What does unpopulate mean?

danuluma replied:
Its an action type I just created. It sets the prePopulated flag to false so that the concept can be edited - after its has mapping has been removed.

dkayiwa commented on src/tests/dictionaryConcepts/container/SelectAnswers.test.js:
```diff
@@ -72,4 +78,19 @@ describe('<SelectAnswers />', () => {
     expect(spy).toHaveBeenCalled();
   });
+
+  it('should handle onClick event', () => {
```
What should handle the on click event?

danuluma replied:
Updated

dkayiwa commented on src/tests/dictionaryConcepts/container/EditConcept.test.jsx:
```diff
@@ -625,4 +652,26 @@ describe('Test suite for mappings on existing concepts', () => {
     expect(instance.state.mappings[1].map_type).toEqual('SAME-AS');
   });
+
+  it('should call removeCurrentAnswer', () => {
```
What should call remove current answer? And when?

danuluma replied:
Updated too

dkayiwa commented on src/components/dictionaryConcepts/containers/EditConcept.jsx:
```diff
@@ -150,9 +157,9 @@ export class EditConcept extends Component {
     event.preventDefault();
   }
 
-  removeNewName(event, uuid) {
```
I thought you changed from uid to uuid?

dkayiwa commented on src/tests/dictionaryConcepts/container/EditConcept.test.jsx:
```diff
@@ -625,4 +652,15 @@ describe('Test suite for mappings on existing concepts', () => {
+    const info = { answerUrl, frontEndUniqueKey, answer };
+    const instance = wrapper.find('EditConcept').instance();
+    const spy = jest.spyOn(instance, 'removeCurrentAnswer');
+    instance.removeCurrentAnswer(info);
```
Are you testing for a method call that you are directly making here?

danuluma replied:
Updated

dkayiwa commented on src/tests/dictionaryConcepts/container/SelectAnswers.test.js:
```diff
@@ -72,4 +88,51 @@ describe('<SelectAnswers />', () => {
+    wrapper = mount(<SelectAnswers {...newProps} />);
+    const event = { key: 'a' };
+    wrapper.find('#searchInputCiel').simulate('keyDown', event);
+    expect(setTimeout).toHaveBeenCalled();
```
How is the above asserting that should clear input and update with current value when a user presses any key?

danuluma replied:
Updated

dkayiwa commented on src/tests/dictionaryConcepts/container/SelectAnswers.test.js:
```diff
@@ -72,4 +88,51 @@ describe('<SelectAnswers />', () => {
+    const instance = wrapper.instance();
+    const e = { key: 'backspace' };
+    instance.resetInput(e);
+    expect(setTimeout).toHaveBeenCalled();
```
How is the above asserting that should clear input when a user presses any non-character key?

danuluma replied:
Updated

dkayiwa commented:
Did you see the merge conflicts on this?

danuluma commented:
Yes, I rebased from master and fixed that. I haven't pushed the latest
changes though. I still working on it.

On Mon, 6 May 2019, 16:55 dkayiwa, <notifications@github.com> wrote:

> Did you see the merge conflicts on this?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openmrs/openmrs-ocl-client/pull/406#issuecomment-489629305>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AIMMXAHH57Y7WJEPVD4IATDPUA2EZANCNFSM4HHC6RTA>
> .
>

alexkayabula commented:
LGTM
